File: mac-app/QuartzTeachers/Models/CourseConfiguration.swift
import Foundation
import Observation

/// An in-memory copy of one course's `course_config.json`.
///
/// This type deliberately stores the decoded JSON as a dictionary and exposes
/// typed accessors over it, rather than using `Codable`. The command-line
/// scripts own this file format; keeping unknown keys intact means a config
/// written by a newer script version survives a round trip through this app.
@Observable
class CourseConfiguration {

    // MARK: - Stored properties

    /// The decoded contents of `course_config.json`.
    var values: [String: Any]

    /// The bytes most recently read from or written to disk, used by
    /// `discardChanges()` to implement the Cancel button.
    private var lastSavedData: Data

    // MARK: - Computed properties

    var courseCode: String {
        return stringValue(forKey: "course_code")
    }

    var courseName: String {
        get { return stringValue(forKey: "course_name") }
        set { values["course_name"] = newValue }
    }

    /// Where deploys go: "netlify" (the default) or "local_folder" — a
    /// folder on this computer, for teachers who upload to their own web
    /// host themselves (e.g. over SFTP). Course-level: every section of
    /// the course publishes the same way.
    var deployTarget: String {
        get {
            let stored: String = stringValue(forKey: "deploy_target")
            return stored.isEmpty ? "netlify" : stored
        }
        set { values["deploy_target"] = newValue }
    }

    /// The folder local-folder deploys publish into; each section lands
    /// in its own sectionN subfolder inside it.
    var deployFolderPath: String {
        get { return stringValue(forKey: "deploy_folder_path") }
        set { values["deploy_folder_path"] = newValue }
    }

    /// True when this course publishes to a folder rather than Netlify.
    var deploysToLocalFolder: Bool {
        return deployTarget == "local_folder" && !deployFolderPath.trimmingCharacters(in: .whitespaces).isEmpty
    }

    /// What is wrong with a folder chosen for local-folder publishing, or
    /// nil when the folder is usable. Both the settings form and the
    /// wizard check this live — and block saving — so a deploy never
    /// discovers the problem after the fact.
    static func deployFolderProblem(forPath rawPath: String) -> String? {
        let path: String = rawPath.trimmingCharacters(in: .whitespaces)
        if path.isEmpty {
            return "Choose the folder this course publishes into."
        }
        let fileManager: FileManager = FileManager.default
        var isDirectory: ObjCBool = false
        let exists: Bool = fileManager.fileExists(atPath: path, isDirectory: &isDirectory)
        if !exists {
            return "That folder doesn’t exist — use Choose… to pick or create one."
        }
        if !isDirectory.boolValue {
            return "That’s a file — publishing needs a folder."
        }
        if !fileManager.isWritableFile(atPath: path) {
            return "That folder can’t be written to — choose a different one."
        }
        return nil
    }

    var customShortName: String {
        get { return stringValue(forKey: "custom_short_name") }
        set { values["custom_short_name"] = newValue }
    }

    var locale: String {
        get {
            let stored: String = stringValue(forKey: "locale")
            if stored.isEmpty {
                return "en-US"
            }
            return stored
        }
        set { values["locale"] = newValue }
    }

    /// True when the course code has no numeric grade in position four
    /// (e.g. "CODING") — the scripts treat these as clubs.
    var isClub: Bool {
        let code: String = courseCode
        if code.count < 4 {
            return true
        }
        let characters: [Character] = Array(code)
        return !characters[3].isNumber
    }

    var sectionNumbers: [Int] {
        if let numbers = values["section_numbers"] as? [Int] {
            return numbers
        }
        // Older configs may store numbers as NSNumber via JSONSerialization.
        if let rawNumbers = values["section_numbers"] as? [Any] {
            var result: [Int] = []
            for rawNumber in rawNumbers {
                if let number = rawNumber as? NSNumber {
                    result.append(number.intValue)
                }
            }
            if !result.isEmpty {
                return result
            }
        }
        let count: Int = intValue(forKey: "num_sections", fallback: 1)
        var result: [Int] = []
        for sectionNumber in 1...max(count, 1) {
            result.append(sectionNumber)
        }
        return result
    }

    var sharedFolders: [String] {
        get { return stringListValue(forKey: "shared_folders") }
        set { values["shared_folders"] = newValue }
    }

    var sharedFiles: [String] {
        get { return stringListValue(forKey: "shared_files") }
        set { values["shared_files"] = newValue }
    }

    var perSectionFolders: [String] {
        get { return stringListValue(forKey: "per_section_folders") }
        set { values["per_section_folders"] = newValue }
    }

    var perSectionFiles: [String] {
        get { return stringListValue(forKey: "per_section_files") }
        set { values["per_section_files"] = newValue }
    }

    var hiddenItems: [String] {
        get { return stringListValue(forKey: "hidden") }
        set { values["hidden"] = newValue }
    }

    var expandableItems: [String] {
        get { return stringListValue(forKey: "expandable") }
        set { values["expandable"] = newValue }
    }

    var expandOnFolderClick: Bool {
        get { return boolValue(forKey: "expandOnFolderClick", fallback: false) }
        set { values["expandOnFolderClick"] = newValue }
    }

    var footerHTML: String {
        get { return stringValue(forKey: "footer_html") }
        set { values["footer_html"] = newValue }
    }

    /// Whether a section's site title leads with the grade ("Grade 12
    /// Computer Science…") — per section, like the section marker. On by
    /// default; the build recomputes the landing title on every build.
    /// An older config that stored one course-wide Bool is honoured.
    func showsGradeInTitle(forSection sectionNumber: Int) -> Bool {
        if let legacy = values["show_grade_in_title"] as? Bool {
            return legacy
        }
        let sectionsMap: [String: Any] = nestedDictionary(forKey: "show_grade_in_title", childKey: "sections")
        if let stored = sectionsMap["section\(sectionNumber)"] as? Bool {
            return stored
        }
        return true
    }

    func setShowsGradeInTitle(_ shows: Bool, forSection sectionNumber: Int) {
        // A legacy course-wide Bool would shadow the per-section map, so
        // it is replaced by the map the first time a section is set.
        if values["show_grade_in_title"] is Bool {
            values["show_grade_in_title"] = [String: Any]()
        }
        setNestedValue(shows, forKey: "show_grade_in_title", childKey: "sections", entryKey: "section\(sectionNumber)")
    }

    /// The landing page's title as the build will compute it — the same
    /// literal rule as `computed_landing_title` in `scripts/build_site.py`:
    /// the grade prefix when its switch is on and the code carries a grade
    /// digit, the course name, and ", Section N" when the marker switch is
    /// on. Used wherever the app previews the site's own headline.
    static func landingTitle(
        courseName: String,
        courseCode: String,
        showsGrade: Bool,
        showsSectionMarker: Bool,
        sectionNumber: Int
    ) -> String {
        let name: String = courseName.trimmingCharacters(in: .whitespaces)
        var prefix: String = ""
        let gradeLabel: String = SectionAdder.gradeLabel(forCourseCode: courseCode)
        if showsGrade && !gradeLabel.isEmpty {
            prefix = "\(gradeLabel) "
        }
        var title: String = "\(prefix)\(name)"
        if showsSectionMarker {
            title = "\(title), Section \(sectionNumber)"
        }
        return title
    }

    /// A quiet warning when turning the grade on would repeat a grade the
    /// course name already carries — "Computer Science, Grade 12, U" would
    /// become "Grade 12 Computer Science, Grade 12, U". The behaviour
    /// stays exactly what the switch says; this only points the problem
    /// out, and the teacher decides: edit the name, or turn the switch off.
    static func gradeInTitleWarning(courseName: String, courseCode: String, showsGrade: Bool) -> String? {
        guard showsGrade else {
            return nil
        }
        let label: String = SectionAdder.gradeLabel(forCourseCode: courseCode)
        guard !label.isEmpty, courseName.contains(label) else {
            return nil
        }
        return "The course name already includes “\(label)”, so the site title would repeat it. Edit the name, or turn this off."
    }

    var showReadingTime: Bool {
        get { return boolValue(forKey: "show_reading_time", fallback: false) }
        set { values["show_reading_time"] = newValue }
    }

    /// Whether this course publishes the Curriculum Coverage page.
    var includesCurriculumCoverage: Bool {
        get { return boolValue(forKey: "include_curriculum_coverage", fallback: true) }
        set {
            values["include_curriculum_coverage"] = newValue
            // The sections cannot outlive the page they sit on.
            if !newValue {
                values["include_coverage_notes"] = false
            }
        }
    }

    /// Whether the coverage page carries its two explanatory sections.
    var includesCoverageNotes: Bool {
        get {
            return CourseConfiguration.coverageNotesEnabled(
                curriculumCoverageEnabled: includesCurriculumCoverage,
                includesCoverageNotes: boolValue(forKey: "include_coverage_notes", fallback: true)
            )
        }
        set { values["include_coverage_notes"] = newValue }
    }

    /// Every item that can be hidden or made expandable in the sidebar.
    var allSidebarItems: [String] {
        var result: [String] = []
        for folder in sharedFolders {
            result.append(folder)
        }
        for file in sharedFiles {
            result.append(file)
        }
        for folder in perSectionFolders {
            result.append(folder)
        }
        for file in perSectionFiles {
            result.append(file)
        }
        return result
    }

    // MARK: - Initializer

    init(values: [String: Any], lastSavedData: Data) {
        self.values = values
        self.lastSavedData = lastSavedData
    }

    /// Loads a configuration from `course_config.json` on disk.
    convenience init(contentsOf url: URL) throws {
        let data: Data = try Data(contentsOf: url)
        let decoded: Any = try JSONSerialization.jsonObject(with: data)
        guard let dictionary = decoded as? [String: Any] else {
            throw CourseConfigurationError.notADictionary
        }
        self.init(values: dictionary, lastSavedData: data)
    }

    // MARK: - Functions

    /// Replaces the course's timetable section numbers (used when a
    /// section is archived and removed).
    func setSectionNumbers(_ sectionNumbers: [Int]) {
        values["section_numbers"] = sectionNumbers
        values["num_sections"] = sectionNumbers.count
    }

    /// The emoji shown beside the site title for a given section.
    func emoji(forSection sectionNumber: Int) -> String {
        let sectionsMap: [String: Any] = nestedDictionary(forKey: "emojis", childKey: "sections")
        if let stored = sectionsMap["section\(sectionNumber)"] as? String {
            if !stored.isEmpty {
                return stored
            }
        }
        return "📚"
    }

    func setEmoji(_ emoji: String, forSection sectionNumber: Int) {
        setNestedValue(emoji, forKey: "emojis", childKey: "sections", entryKey: "section\(sectionNumber)")
    }

    /// The teacher's own domain for a section's published site — shown in
    /// links to the live site in place of the address Netlify assigns.
    /// Empty when the Netlify address is used as-is.
    func customDomain(forSection sectionNumber: Int) -> String {
        let sectionsMap: [String: Any] = nestedDictionary(forKey: "custom_domains", childKey: "sections")
        if let stored = sectionsMap["section\(sectionNumber)"] as? String {
            return stored
        }
        return ""
    }

    func setCustomDomain(_ domain: String, forSection sectionNumber: Int) {
        setNestedValue(domain, forKey: "custom_domains", childKey: "sections", entryKey: "section\(sectionNumber)")
    }

    /// A typed or pasted domain, reduced to just the domain: whitespace
    /// trimmed, any scheme stripped, and anything from the first slash on
    /// dropped — so a pasted "https://ics3u.school.ca/" stores as
    /// "ics3u.school.ca".
    /// Whether the curriculum coverage map should be switched on for a new
    /// course.
    ///
    /// The map is drawn from the site's own links to the curriculum pages,
    /// so it cannot exist without them: declining the curriculum forces the
    /// map off, whatever the toggle was last left at. The reverse is not
    /// true — keeping the curriculum pages and declining the map is a
    /// reasonable choice, and this returns exactly what the teacher asked
    /// for in that case.
    ///
    /// The rule lives here rather than inside the wizard view so that it
    /// can be tested: a SwiftUI `@State` property has no backing store
    /// until the view is on screen, so a test that sets one and reads a
    /// computed result gets the default back every time.
    static func curriculumCoverageEnabled(
        codeHasExampleContent: Bool,
        prepopulatesExampleContent: Bool,
        payloadIncludesCurriculum: Bool,
        includesCurriculumPages: Bool,
        includesCurriculumCoverage: Bool
    ) -> Bool {
        guard codeHasExampleContent,
              prepopulatesExampleContent,
              payloadIncludesCurriculum,
              includesCurriculumPages else {
            return false
        }
        return includesCurriculumCoverage
    }

    /// Whether the coverage page's two explanatory sections should be
    /// switched on for a new course.
    ///
    /// The sections live on the coverage page, so they cannot exist without
    /// it: declining the map forces them off whatever the toggle was last
    /// left at. Keeping the map and declining the sections is a reasonable
    /// choice — a page published to students is often better as the map
    /// alone — and this returns exactly that.
    static func coverageNotesEnabled(
        curriculumCoverageEnabled: Bool,
        includesCoverageNotes: Bool
    ) -> Bool {
        guard curriculumCoverageEnabled else {
            return false
        }
        return includesCoverageNotes
    }

    static func normalizedCustomDomain(_ raw: String) -> String {
        var domain: String = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        for scheme in ["https://", "http://"] {
            if domain.hasPrefix(scheme) {
                domain = String(domain.dropFirst(scheme.count))
            }
        }
        if let slashIndex = domain.firstIndex(of: "/") {
            domain = String(domain[..<slashIndex])
        }
        return domain
    }

    /// Whether the site title shows the "S1"-style marker for a given section.
    func showsSectionMarker(forSection sectionNumber: Int) -> Bool {
        let sectionsMap: [String: Any] = nestedDictionary(forKey: "show_section_marker", childKey: "sections")
        if let stored = sectionsMap["section\(sectionNumber)"] as? Bool {
            return stored
        }
        return true
    }

    func setShowsSectionMarker(_ shows: Bool, forSection sectionNumber: Int) {
        setNestedValue(shows, forKey: "show_section_marker", childKey: "sections", entryKey: "section\(sectionNumber)")
    }

    /// The colour scheme id chosen for a given section, or "" when unset.
    func colourSchemeID(forSection sectionNumber: Int) -> String {
        if let map = values["color_schemes"] as? [String: Any] {
            if let stored = map["section\(sectionNumber)"] as? String {
                return stored
            }
        }
        return ""
    }

    func setColourSchemeID(_ schemeID: String, forSection sectionNumber: Int) {
        var map: [String: Any] = [:]
        if let existing = values["color_schemes"] as? [String: Any] {
            map = existing
        }
        map["section\(sectionNumber)"] = schemeID
        values["color_schemes"] = map
    }

    /// The font choices for a section, falling back to the course default.
    func fontChoice(forSection sectionNumber: Int) -> FontChoice {
        let sectionsMap: [String: Any] = nestedDictionary(forKey: "fonts", childKey: "sections")
        if let stored = sectionsMap["section\(sectionNumber)"] as? [String: Any] {
            return FontChoice(dictionary: stored)
        }
        if let fonts = values["fonts"] as? [String: Any] {
            if let defaultChoice = fonts["default"] as? [String: Any] {
                return FontChoice(dictionary: defaultChoice)
            }
        }
        return FontChoice.systemDefault
    }

    func setFontChoice(_ choice: FontChoice, forSection sectionNumber: Int) {
        setNestedValue(choice.dictionaryRepresentation, forKey: "fonts", childKey: "sections", entryKey: "section\(sectionNumber)")
    }

    var defaultFontChoice: FontChoice {
        get {
            if let fonts = values["fonts"] as? [String: Any] {
                if let stored = fonts["default"] as? [String: Any] {
                    return FontChoice(dictionary: stored)
                }
            }
            return FontChoice.systemDefault
        }
        set {
            var fonts: [String: Any] = [:]
            if let existing = values["fonts"] as? [String: Any] {
                fonts = existing
            }
            fonts["default"] = newValue.dictionaryRepresentation
            values["fonts"] = fonts
        }
    }

    /// Writes the configuration to disk in the same shape the setup wizard
    /// uses: pretty-printed JSON with a trailing newline.
    func write(to url: URL) throws {
        let options: JSONSerialization.WritingOptions = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        var data: Data = try JSONSerialization.data(withJSONObject: values, options: options)
        data.append(contentsOf: [0x0A])
        try data.write(to: url, options: [.atomic])
        lastSavedData = data
    }

    /// Reverts all in-memory edits back to the last data read from or
    /// written to disk (the Cancel button).
    func discardChanges() throws {
        let decoded: Any = try JSONSerialization.jsonObject(with: lastSavedData)
        guard let dictionary = decoded as? [String: Any] else {
            throw CourseConfigurationError.notADictionary
        }
        values = dictionary
    }

    /// True when the in-memory values differ from what was last saved.
    var hasUnsavedChanges: Bool {
        guard let savedDecoded = try? JSONSerialization.jsonObject(with: lastSavedData) else {
            return true
        }
        guard let savedDictionary = savedDecoded as? [String: Any] else {
            return true
        }
        let current = values as NSDictionary
        let saved = savedDictionary as NSDictionary
        return !current.isEqual(to: saved as! [AnyHashable: Any])
    }

    // MARK: - Private helpers

    private func stringValue(forKey key: String) -> String {
        if let stored = values[key] as? String {
            return stored
        }
        return ""
    }

    private func intValue(forKey key: String, fallback: Int) -> Int {
        if let stored = values[key] as? NSNumber {
            return stored.intValue
        }
        return fallback
    }

    private func boolValue(forKey key: String, fallback: Bool) -> Bool {
        if let stored = values[key] as? Bool {
            return stored
        }
        return fallback
    }

    private func stringListValue(forKey key: String) -> [String] {
        if let stored = values[key] as? [String] {
            return stored
        }
        return []
    }

    private func nestedDictionary(forKey key: String, childKey: String) -> [String: Any] {
        if let outer = values[key] as? [String: Any] {
            if let inner = outer[childKey] as? [String: Any] {
                return inner
            }
        }
        return [:]
    }

    private func setNestedValue(_ value: Any, forKey key: String, childKey: String, entryKey: String) {
        var outer: [String: Any] = [:]
        if let existing = values[key] as? [String: Any] {
            outer = existing
        }
        var inner: [String: Any] = [:]
        if let existingInner = outer[childKey] as? [String: Any] {
            inner = existingInner
        }
        inner[entryKey] = value
        outer[childKey] = inner
        values[key] = outer
    }
}

enum CourseConfigurationError: Error {
    case notADictionary
}

File: windows-app/Plantoir.Core/Models/CourseConfiguration.cs
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Plantoir.Core.Models;

/// <summary>
/// The in-memory copy of one course's course_config.json.
///
/// Deliberately NOT a typed model: the decoded JSON lives in a mutable
/// JObject and edits happen key by key, so keys written by a newer
/// toolchain version survive a settings round trip untouched. The CLI
/// scripts own the format; this class is a careful guest in their file.
///
/// Written form matches what the toolchain leaves on disk: 2-space indent,
/// keys sorted ascending (ordinal), raw non-ASCII (emoji stay emoji),
/// LF newlines, one trailing newline.
/// </summary>
public sealed class CourseConfiguration
{
    private JObject _values;
    private byte[] _lastSavedData;

    private CourseConfiguration(JObject values, byte[] lastSavedData)
    {
        _values = values;
        _lastSavedData = lastSavedData;
    }

    public static CourseConfiguration Load(string path) => FromBytes(File.ReadAllBytes(path));

    public static CourseConfiguration FromBytes(byte[] data)
    {
        var token = JToken.Parse(Encoding.UTF8.GetString(data));
        if (token is not JObject obj)
            throw new InvalidDataException("course_config.json does not hold a JSON object.");
        return new CourseConfiguration(obj, data);
    }

    public static CourseConfiguration FromDictionary(JObject values)
    {
        var config = new CourseConfiguration(values, Array.Empty<byte>());
        config._lastSavedData = config.SerializedBytes();
        return config;
    }

    // ---- Round-trip contract -------------------------------------------

    public byte[] SerializedBytes()
    {
        var text = new StringWriter { NewLine = "\n" };
        using (var writer = new JsonTextWriter(text)
        {
            Formatting = Formatting.Indented,
            Indentation = 2,
            IndentChar = ' ',
        })
        {
            SortedCopy(_values).WriteTo(writer);
        }
        return Encoding.UTF8.GetBytes(text.ToString() + "\n");
    }

    private static JToken SortedCopy(JToken node) => node switch
    {
        JObject obj => new JObject(obj.Properties()
            .OrderBy(p => p.Name, StringComparer.Ordinal)
            .Select(p => new JProperty(p.Name, SortedCopy(p.Value)))),
        JArray array => new JArray(array.Select(SortedCopy)),
        _ => node.DeepClone(),
    };

    public void Write(string path)
    {
        byte[] data = SerializedBytes();
        string temp = path + ".tmp";
        File.WriteAllBytes(temp, data);
        File.Move(temp, path, overwrite: true);
        _lastSavedData = data;
    }

    /// <summary>The Revert button: put the values back the way the last save left them.</summary>
    public void DiscardChanges()
    {
        if (_lastSavedData.Length == 0) return;
        if (JToken.Parse(Encoding.UTF8.GetString(_lastSavedData)) is JObject obj) _values = obj;
    }

    public bool HasUnsavedChanges
    {
        get
        {
            if (_lastSavedData.Length == 0) return true;
            try
            {
                var saved = JToken.Parse(Encoding.UTF8.GetString(_lastSavedData));
                return !JToken.DeepEquals(SortedCopy(_values), SortedCopy(saved));
            }
            catch { return true; }
        }
    }

    public JObject Values => _values;

    // ---- Flat keys ------------------------------------------------------

    public string CourseCode => StringValue("course_code");

    public string CourseName
    {
        get => StringValue("course_name");
        set => _values["course_name"] = value;
    }

    public string CustomShortName
    {
        get => StringValue("custom_short_name");
        set => _values["custom_short_name"] = value;
    }

    /// <summary>
    /// Where publishes go: "netlify" (the default) or "local_folder" — a
    /// folder on this computer, for teachers who upload to their own web
    /// host themselves (e.g. over SFTP). Course-level: every section of the
    /// course publishes the same way.
    /// </summary>
    public string DeployTarget
    {
        get { var raw = StringValue("deploy_target"); return raw.Length == 0 ? "netlify" : raw; }
        set => _values["deploy_target"] = value;
    }

    /// <summary>
    /// The folder local-folder publishes land in; each section gets its own
    /// sectionN subfolder inside it.
    /// </summary>
    public string DeployFolderPath
    {
        get => StringValue("deploy_folder_path");
        set => _values["deploy_folder_path"] = value;
    }

    /// <summary>True when this course publishes to a folder rather than Netlify.</summary>
    public bool DeploysToLocalFolder =>
        DeployTarget == "local_folder" && DeployFolderPath.Trim().Length > 0;

    /// <summary>True when this course publishes to Cloudflare Pages.</summary>
    public bool DeploysToCloudflare => DeployTarget == "cloudflare_pages";

    /// <summary>
    /// What is wrong with a Cloudflare account ID, or null when it looks
    /// usable. A token scoped to Pages cannot list its own account — verified
    /// against a real token — so the teacher supplies it once here, and the
    /// app checks the shape before a publish can fail on it.
    /// </summary>
    public static string? CloudflareAccountProblem(string rawId)
    {
        string id = rawId.Trim();
        if (id.Length == 0) return "Paste your Cloudflare Account ID.";
        if (id.Length != 32 || !id.All(Uri.IsHexDigit))
            return "That doesn’t look like an Account ID — it’s 32 letters and digits.";
        return null;
    }

    /// <summary>
    /// What is wrong with a folder chosen for local-folder publishing, or
    /// null when the folder is usable. Both the settings form and the wizard
    /// check this live — and block saving — so a publish never discovers the
    /// problem after the fact.
    /// </summary>
    public static string? DeployFolderProblem(string rawPath)
    {
        string path = rawPath.Trim();
        if (path.Length == 0) return "Choose the folder this course deploys into.";
        if (File.Exists(path)) return "That’s a file — publishing needs a folder.";
        if (!Directory.Exists(path)) return "That folder doesn’t exist — use Choose… to pick or create one.";
        try
        {
            string probe = Path.Combine(path, ".plantoir-write-probe-" + Guid.NewGuid().ToString("N"));
            File.WriteAllText(probe, "");
            File.Delete(probe);
        }
        catch
        {
            return "That folder can’t be written to — choose a different one.";
        }
        return null;
    }

    public string Locale
    {
        get { var raw = StringValue("locale"); return raw.Length == 0 ? "en-US" : raw; }
        set => _values["locale"] = value;
    }

    public bool ExpandOnFolderClick
    {
        get => BoolValue("expandOnFolderClick", false);
        set => _values["expandOnFolderClick"] = value;
    }

    public bool ShowReadingTime
    {
        get => BoolValue("show_reading_time", false);
        set => _values["show_reading_time"] = value;
    }

    public string FooterHtml
    {
        get => StringValue("footer_html");
        set => _values["footer_html"] = value;
    }

    public List<string> SharedFolders { get => StringList("shared_folders"); set => SetStringList("shared_folders", value); }
    public List<string> SharedFiles { get => StringList("shared_files"); set => SetStringList("shared_files", value); }
    public List<string> PerSectionFolders { get => StringList("per_section_folders"); set => SetStringList("per_section_folders", value); }
    public List<string> PerSectionFiles { get => StringList("per_section_files"); set => SetStringList("per_section_files", value); }
    public List<string> HiddenItems { get => StringList("hidden"); set => SetStringList("hidden", value); }
    public List<string> ExpandableItems { get => StringList("expandable"); set => SetStringList("expandable", value); }

    /// <summary>sharedFolders + sharedFiles + perSectionFolders + perSectionFiles, in that order.</summary>
    public List<string> AllSidebarItems =>
        SharedFolders.Concat(SharedFiles).Concat(PerSectionFolders).Concat(PerSectionFiles).ToList();

    /// <summary>A club has a short code, or a non-digit where a course carries its grade.</summary>
    public bool IsClub
    {
        get
        {
            string code = CourseCode;
            if (code.Length < 4) return true;
            return !char.IsDigit(code[3]);
        }
    }

    // ---- Section numbers (matching the mac app's three tiers) -----------

    public List<int> SectionNumbers
    {
        get
        {
            if (_values["section_numbers"] is JArray array)
            {
                // An empty stored list IS the answer (a course whose every
                // section was archived); only an unreadable list falls through.
                var parsed = array
                    .Where(t => t.Type is JTokenType.Integer or JTokenType.Float)
                    .Select(t => (int)t)
                    .ToList();
                if (array.Count == 0 || parsed.Count > 0) return parsed;
            }
            int count = IntValue("num_sections", 1);
            return Enumerable.Range(1, Math.Max(count, 1)).ToList();
        }
    }

    /// <summary>Writes BOTH section_numbers and num_sections = count.</summary>
    public void SetSectionNumbers(IReadOnlyList<int> numbers)
    {
        _values["section_numbers"] = new JArray(numbers);
        _values["num_sections"] = numbers.Count;
    }

    // ---- Per-section nested maps ({key: {"sections": {"sectionN": v}}}) --

    public string Emoji(int section)
    {
        string raw = NestedString("emojis", "sections", SectionKey(section));
        return raw.Length == 0 ? "📚" : raw;
    }

    public void SetEmoji(int section, string emoji) =>
        SetNestedValue("emojis", "sections", SectionKey(section), emoji);

    public string CustomDomain(int section) =>
        NestedString("custom_domains", "sections", SectionKey(section));

    public void SetCustomDomain(int section, string domain) =>
        SetNestedValue("custom_domains", "sections", SectionKey(section), NormalizedCustomDomain(domain));

    public bool ShowsSectionMarker(int section) =>
        NestedBool("show_section_marker", "sections", SectionKey(section), true);

    public void SetShowsSectionMarker(int section, bool value) =>
        SetNestedValue("show_section_marker", "sections", SectionKey(section), value);

    /// <summary>
    /// Legacy rule: a plain course-wide Bool wins for every section until the
    /// first per-section write replaces it with a map.
    /// </summary>
    public bool ShowsGradeInTitle(int section)
    {
        if (_values["show_grade_in_title"] is JValue { Type: JTokenType.Boolean } legacy)
            return (bool)legacy;
        return NestedBool("show_grade_in_title", "sections", SectionKey(section), true);
    }

    public void SetShowsGradeInTitle(int section, bool value)
    {
        if (_values["show_grade_in_title"] is JValue { Type: JTokenType.Boolean })
            _values["show_grade_in_title"] = new JObject();
        SetNestedValue("show_grade_in_title", "sections", SectionKey(section), value);
    }

    /// <summary>color_schemes is FLAT — {"color_schemes": {"sectionN": "id"}}, no "sections" wrapper.</summary>
    public string ColourSchemeId(int section) =>
        _values["color_schemes"] is JObject map && map[SectionKey(section)] is JValue { Type: JTokenType.String } v
            ? (string)v! : "";

    public void SetColourSchemeId(int section, string id)
    {
        if (_values["color_schemes"] is not JObject map)
        {
            map = new JObject();
            _values["color_schemes"] = map;
        }
        map[SectionKey(section)] = id;
    }

    /// <summary>fonts.sections.sectionN → fonts.default → system default.</summary>
    public FontChoice Font(int section)
    {
        if (_values["fonts"] is JObject fonts)
        {
            if (fonts["sections"] is JObject sections && sections[SectionKey(section)] is JObject perSection)
                return FontChoice.FromJson(perSection);
            if (fonts["default"] is JObject fallback)
                return FontChoice.FromJson(fallback);
        }
        return FontChoice.SystemDefault;
    }

    public void SetFont(int section, FontChoice choice)
    {
        if (_values["fonts"] is not JObject fonts) { fonts = new JObject(); _values["fonts"] = fonts; }
        if (fonts["sections"] is not JObject sections) { sections = new JObject(); fonts["sections"] = sections; }
        sections[SectionKey(section)] = choice.ToJson();
    }

    // ---- Rules ----------------------------------------------------------

    /// <summary>Trim → strip a leading https:// then http:// → cut at the first slash.</summary>
    public static string NormalizedCustomDomain(string raw)
    {
        string s = raw.Trim();
        if (s.StartsWith("https://", StringComparison.Ordinal)) s = s["https://".Length..];
        else if (s.StartsWith("http://", StringComparison.Ordinal)) s = s["http://".Length..];
        int slash = s.IndexOf('/');
        if (slash >= 0) s = s[..slash];
        return s;
    }

    /// <summary>
    /// The quiet orange warning: the switch stays literal, but when it is on
    /// and the course name already carries its grade label, say so.
    /// </summary>
    public static string? GradeInTitleWarning(string courseName, string courseCode, bool showsGrade)
    {
        if (!showsGrade) return null;
        string label = SectionAdder.GradeLabel(courseCode);
        if (label.Length == 0 || !courseName.Contains(label, StringComparison.Ordinal)) return null;
        return $"The course name already includes “{label}”, so the site title would repeat it. Edit the name, or turn this off.";
    }

    /// <summary>
    /// The site title as build_site.py's computed_landing_title will compute
    /// it at build time — [Grade X ]Name[, Section N], deliberately literal:
    /// the grade switch alone decides the prefix, the marker switch alone
    /// decides the suffix, and an empty name falls back to the code.
    /// </summary>
    public static string ComputedSiteTitle(string courseName, string courseCode, int sectionNumber,
                                           bool showsGrade, bool showsMarker)
    {
        string code = courseCode.Trim().ToUpperInvariant();
        string name = courseName.Trim();
        if (name.Length == 0) name = code;
        string label = SectionAdder.GradeLabel(code);
        string prefix = showsGrade && label.Length > 0 ? label + " " : "";
        string suffix = showsMarker ? $", Section {sectionNumber}" : "";
        return prefix + name + suffix;
    }

    // ---- Helpers --------------------------------------------------------

    internal static string SectionKey(int section) => "section" + section;

    private string StringValue(string key) =>
        _values[key] is JValue { Type: JTokenType.String } v ? (string)v! : "";

    private bool BoolValue(string key, bool fallback) =>
        _values[key] is JValue { Type: JTokenType.Boolean } v ? (bool)v : fallback;

    private int IntValue(string key, int fallback) =>
        _values[key] is JValue { Type: JTokenType.Integer } v ? (int)v : fallback;

    private List<string> StringList(string key)
    {
        var result = new List<string>();
        if (_values[key] is JArray array)
            foreach (var element in array)
                if (element is JValue { Type: JTokenType.String } v) result.Add((string)v!);
        return result;
    }

    private void SetStringList(string key, IReadOnlyList<string> items) =>
        _values[key] = new JArray(items);

    private string NestedString(string key, string childKey, string entryKey) =>
        _values[key] is JObject outer && outer[childKey] is JObject inner &&
        inner[entryKey] is JValue { Type: JTokenType.String } v ? (string)v! : "";

    private bool NestedBool(string key, string childKey, string entryKey, bool fallback) =>
        _values[key] is JObject outer && outer[childKey] is JObject inner &&
        inner[entryKey] is JValue { Type: JTokenType.Boolean } v ? (bool)v : fallback;

    private void SetNestedValue(string key, string childKey, string entryKey, JToken value)
    {
        if (_values[key] is not JObject outer) { outer = new JObject(); _values[key] = outer; }
        if (outer[childKey] is not JObject inner) { inner = new JObject(); outer[childKey] = inner; }
        inner[entryKey] = value;
    }
}

/// <summary>Header, body, and code font display names for one section.</summary>
public readonly record struct FontChoice(string Header, string Body, string Code)
{
    public static FontChoice SystemDefault { get; } = new("Helvetica, Arial", "Helvetica, Arial", "IBM Plex Mono");

    public static FontChoice FromJson(JObject obj) => new(
        Field(obj, "header", SystemDefault.Header),
        Field(obj, "body", SystemDefault.Body),
        Field(obj, "code", SystemDefault.Code));

    private static string Field(JObject obj, string key, string fallback) =>
        obj[key] is JValue { Type: JTokenType.String } v ? (string)v! : fallback;

    public JObject ToJson() => new()
    {
        ["header"] = Header,
        ["body"] = Body,
        ["code"] = Code,
    };
}
